resteasy-jaxrs/src/main/java/org/jboss/resteasy/spi/ResteasyProviderFactory.java

```diff
@@ -1887,29 +1886,40 @@ public MediaType getConcreteMediaTypeFromMessageBodyWriters(Class type, Type gen
       return null;
    }
    
+   @Deprecated
    public List<MessageBodyWriter<?>> getPossibleMessageBodyWriters(Class type, Type genericType, Annotation[] annotations)
    {
-      List<MessageBodyWriter<?>> list = new ArrayList<MessageBodyWriter<?>>();
+      return new ArrayList<>(getPossibleMessageBodyWritersMap(type, genericType, annotations).keySet());
+   }
+   
+   public Map<MessageBodyWriter<?>, Class<?>> getPossibleMessageBodyWritersMap(Class type, Type genericType, Annotation[] annotations)
+   {
+      Map<MessageBodyWriter<?>, Class<?>> map = new HashMap<MessageBodyWriter<?>, Class<?>>();
       MediaType mediaType = MediaType.WILDCARD_TYPE;
       List<SortedKey<MessageBodyWriter>> writers = getServerMessageBodyWriters().getPossible(mediaType, type);
       for (SortedKey<MessageBodyWriter> writer : writers)
       {
          if (writer.obj.isWriteable(type, genericType, annotations, mediaType))
          {
             MessageBodyWriter mbw = writer.obj;
-            Class writerType = Types.getTemplateParameterOfInterface(mbw.getClass(), MessageBodyWriter.class);
-            if (writerType == null || writerType.equals(Object.class) || !writerType.isAssignableFrom(type)) continue;
+            Class<?> mbwc = mbw.getClass();
+            if (!mbwc.isInterface() && mbwc.getSuperclass() != null && !mbwc.getSuperclass().equals(Object.class) && WeldUtil.isWeldProxy(mbwc)) {
+               mbwc = mbwc.getSuperclass();
+            }
+            Class writerType = Types.getTemplateParameterOfInterface(mbwc, MessageBodyWriter.class);
+            if (writerType == null || !writerType.isAssignableFrom(type)) continue;
             Produces produces = mbw.getClass().getAnnotation(Produces.class);
             if (produces == null) continue;
             for (String produce : produces.value())
             {
                MediaType mt = MediaType.valueOf(produce);
                if (mt.isWildcardType() || mt.isWildcardSubtype()) continue;

```

Comment:
Why are we filtering `MessageBodyWriter ` with produces media types `isWildcardType() || isWildcardSubtype())` ?

Suppose that the client is asking for media type "x/y" and that the only `MessageBodyWriter` able to handle it is the one annotated with `@Produces("x/*")` it will be filtered.

To my mind it should be accepted and then when computing for most specific media types between "x/y" and "x/*" it will be "x/y".